### PR TITLE
Support Snappy compression for gRPC

### LINKED ISSUE #2898: [feat] support snappy compressor for gRPC
### WHAT

Let's support snappy compressor for gRPC.

### WHY

Currently, we only have gzip for gRPC compressor.
The gzip use too much CPU.
The snappy is better than gzip for real-time compressing.

### PR DESCRIPTION
**Pull request by Wing924**
**What this PR does**:

Add snappy compression for gRPC.

**Which issue(s) this PR fixes**:
Fixes #2898

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/grpcclient/grpcclient.go [hunk 1]
```diff
@@ -31,21 +36,40 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.IntVar(&cfg.MaxRecvMsgSize, prefix+".grpc-max-recv-msg-size", 100<<20, "gRPC client max receive message size (bytes).")
 	f.IntVar(&cfg.MaxSendMsgSize, prefix+".grpc-max-send-msg-size", 16<<20, "gRPC client max send message size (bytes).")
-	f.BoolVar(&cfg.UseGzipCompression, prefix+".grpc-use-gzip-compression", false, "Use compression when sending messages.")
+	f.BoolVar(&cfg.UseGzipCompression, prefix+".grpc-use-gzip-compression", false, "Deprecated: Use gzip compression when sending messages.")
+	f.StringVar(&cfg.UseCompression, prefix+".grpc-use-compression", "", "Use compression when sending messages. Supported values are: 'gzip', 'snappy' and '' (disable compression)")
```
Comment (pstibrany) [inline pkg/util/grpcclient/grpcclient.go:40]: I'd suggest to call this just `Compression` and `.grpc-compression`, as "use-compression" sounds like a boolean flag.

File: pkg/util/grpcclient/grpcclient.go [hunk 2]
```diff
@@ -31,21 +36,40 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	f.IntVar(&cfg.RateLimitBurst, prefix+".grpc-client-rate-limit-burst", 0, "Rate limit burst for gRPC client.")
 	f.BoolVar(&cfg.BackoffOnRatelimits, prefix+".backoff-on-ratelimits", false, "Enable backoff and retry when we hit ratelimits.")
 
 	cfg.BackoffConfig.RegisterFlags(prefix, f)
 }
 
+func (cfg *Config) Validate(log log.Logger) error {
+	if cfg.UseGzipCompression {
+		flagext.DeprecatedFlagsUsed.Inc()
+		level.Warn(log).Log("msg", "flag *.grpc-use-gzip-compression (or config use_gzip_compression) is deprecated, use *.grpc-use-compression instead.")
```
Comment (pstibrany) [inline pkg/util/grpcclient/grpcclient.go:51]: Not sure if it's a good idea, but we could save the prefix in RegisterFlagsWithPrefix, and then report full flag name here.

Comment (pstibrany) [inline pkg/util/grpcclient/grpcclient.go:51]: We use `running with DEPRECATED flag ...` message in other places.

File: CHANGELOG.md [hunk 3]
```diff
@@ -68,6 +68,13 @@
 * [ENHANCEMENT] Flusher: Added `-flusher.exit-after-flush` option (defaults to true) to control whether Cortex should stop completely after Flusher has finished its work. #2877
 * [ENHANCEMENT] Added metrics `cortex_config_hash` and `cortex_runtime_config_hash` to expose hash of the currently active config file. #2874
 * [ENHANCEMENT] Logger: added JSON logging support, configured via the `-log.format=json` CLI flag or its respective YAML config option. #2386
+* [ENHANCEMENT] Support Snappy compression for gRPC. #2940
```
Comment (pstibrany) [inline CHANGELOG.md:71]: I would rephrase this slightly as: Added new flags to configure compression used by gRPC (mention flags here). Valid values are "gzip", "snappy", "" (no compression, default). Previous flags (*.use-gzip-compression, enumerate) are now deprecated. WDYT?

Comment (gouthamve) [reply CHANGELOG.md:71]: As we are deprecating things, we should move the deprecations to `CHANGE` as well.

**Comment by bboreham** [comment]
Could you give specific numbers for the benchmarks - looks like a 40% reduction in CPU usage traded for a 150% change in network bandwidth?

**Comment by Wing924** [comment]
### benchmark result

| resource            | none   | gzip    | snappy | none -> gzip | none -> snappy | gzip -> snappy |
| ------------------- | ------ | ------- | ------ | ------------ | -------------- | -------------- |
| distributor cpu     | 23%    | 40%     | 25%    | +74%         | +8%            | -38%           |
| ingester cpu        | 43%    | 46%     | 49%    | +6%          | +13%           | +7%            |
| distributor network | 74Mbps | 8.2Mbps | 12Mbps | -88%         | -83%           | +46%           |
| ingester network    | 93Mbps | 6.5Mbps | 12Mbps | -93%         | -87%           | +84%           |

**Comment by Wing924** [comment]
@pstibrany I've fixed it. Could you review it again?

File: CHANGELOG.md [hunk 4]
```diff
@@ -40,6 +40,13 @@
     - `-experimental.tsdb.stripe-size` changed to `-experimental.blocks-storage.tsdb.stripe-size`
     - `-experimental.tsdb.wal-compression-enabled` changed to `-experimental.blocks-storage.tsdb.wal-compression-enabled`
     - `-experimental.tsdb.flush-blocks-on-shutdown` changed to `-experimental.blocks-storage.tsdb.flush-blocks-on-shutdown`
+* [ENHANCEMENT/CHANGE] Added new flags `-*.grpc-compression` to configure compression used by gRPC. Valid values are "gzip", "snappy", "" (no compression, default). Previous flags (`-*.use-gzip-compression`, enumerate) are now deprecated. #2940
+  * `-bigtable.grpc-compression`
+  * `-ingester.client.grpc-compression`
+  * `-querier.frontend-client.grpc-compression`
+  * Deprecated: `-bigtable.grpc-use-gzip-compression`
+  * Deprecated: `-ingester.client.grpc-use-gzip-compression`
+  * Deprecated: `-querier.frontend-client.grpc-use-gzip-compression`
```
Comment (pstibrany) [inline CHANGELOG.md:49]:
```suggestion
* [ENHANCEMENT] Added new flags `-bigtable.grpc-compression`, `-ingester.client.grpc-compression`, `-querier.frontend-client.grpc-compression` to configure compression used by gRPC. Valid values are `gzip`, `snappy`, or empty string (no compression, default). #2940
* [CHANGE] Flags `-bigtable.grpc-use-gzip-compression`, `-ingester.client.grpc-use-gzip-compression`, `-querier.frontend-client.grpc-use-gzip-compression` are now deprecated. #2940
```

(Note that they should be moved to their respective sections)

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
+	dr := c.readersPool.Get().(*reader)
+	dr.Reset(r)
+	return dr, nil
+}
+
+type writeCloser struct {
+	*compressor
+	*snappy.Writer
```
Comment (pstibrany) [inline pkg/util/grpc/encoding/snappy/snappy.go:62]: I'd prefer to avoid nameless embeds. They expose more methods than intended (writeCloser now has methods like `Name` and `Compress`), and may inadvertently implement methods that were not meant to be implemented.

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
+package snappy
```
Comment (pstibrany) [inline pkg/util/grpc/encoding/snappy/snappy.go:1]: So many nested pkg names feel so "Java"-like. Any downside to having it directly in `pkg/util/grpc/encoding` package?

File: pkg/util/grpcclient/grpcclient.go [hunk 7]
```diff
@@ -29,23 +37,43 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	f.IntVar(&cfg.RateLimitBurst, prefix+".grpc-client-rate-limit-burst", 0, "Rate limit burst for gRPC client.")
 	f.BoolVar(&cfg.BackoffOnRatelimits, prefix+".backoff-on-ratelimits", false, "Enable backoff and retry when we hit ratelimits.")
 
 	cfg.BackoffConfig.RegisterFlags(prefix, f)
 }
 
+func (cfg *Config) Validate(log log.Logger) error {
+	if cfg.UseGzipCompression {
+		flagext.DeprecatedFlagsUsed.Inc()
+		level.Warn(log).Log("msg", fmt.Sprintf("running with DEPRECATED flag -%s.grpc-use-gzip-compression, use -%s.grpc-use-compression instead.", cfg.prefix, cfg.prefix))
```
Comment (pstibrany) [inline pkg/util/grpcclient/grpcclient.go:55]: should be `... use -%s.grpc-compression` now.

File: pkg/ingester/client/client.go [hunk 8]
```diff
@@ -3,12 +3,14 @@ package client
...
 	"flag"
 
+	"github.com/go-kit/kit/log"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promauto"
 	"google.golang.org/grpc"
 	_ "google.golang.org/grpc/encoding/gzip" // get gzip compressor registered
 	"google.golang.org/grpc/health/grpc_health_v1"
 
+	_ "github.com/cortexproject/cortex/pkg/util/grpc/encoding/snappy" // get snappy compressor registered
```
Comment (pstibrany) [inline pkg/ingester/client/client.go:13]: I think we should move import this import into `pkg/util/grpcclient`, where it is used. Importing it only here will not work for code that doesn't use `pkg/ingester/client`, but still uses `pkg/util/grpcclient`.

File: pkg/util/grpcclient/grpcclient.go [hunk 9]
```diff
@@ -29,23 +37,43 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 
 // RegisterFlagsWithPrefix registers flags with prefix.
 func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	cfg.prefix = prefix
 	f.IntVar(&cfg.MaxRecvMsgSize, prefix+".grpc-max-recv-msg-size", 100<<20, "gRPC client max receive message size (bytes).")
 	f.IntVar(&cfg.MaxSendMsgSize, prefix+".grpc-max-send-msg-size", 16<<20, "gRPC client max send message size (bytes).")
-	f.BoolVar(&cfg.UseGzipCompression, prefix+".grpc-use-gzip-compression", false, "Use compression when sending messages.")
+	f.BoolVar(&cfg.UseGzipCompression, prefix+".grpc-use-gzip-compression", false, "Deprecated: Use gzip compression when sending messages.")
```
Comment (pstibrany) [inline pkg/util/grpcclient/grpcclient.go:43]:
```suggestion
	f.BoolVar(&cfg.UseGzipCompression, prefix+".grpc-use-gzip-compression", false, "Deprecated: Use gzip compression when sending messages. If true, overrides -" + prefix + ".grpc-compression flag.")
```

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+// Name is the name registered for the snappy compressor.
+const Name = "snappy"
+
+func init() {
+	encoding.RegisterCompressor(newCompressor())
+}
+
+type compressor struct {
```
Comment (pstibrany) [inline pkg/util/grpc/encoding/snappy/snappy.go:18]: It would be nice to have small unit test for `compressor` type? WDYT?

File: pkg/util/grpcclient/grpcclient.go [hunk 11]
```diff
@@ -29,23 +37,43 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	cfg.BackoffConfig.RegisterFlags(prefix, f)
 }
 
+func (cfg *Config) Validate(log log.Logger) error {
+	if cfg.UseGzipCompression {
+		flagext.DeprecatedFlagsUsed.Inc()
+		level.Warn(log).Log("msg", fmt.Sprintf("running with DEPRECATED flag -%s.grpc-use-gzip-compression, use -%s.grpc-use-compression instead.", cfg.prefix, cfg.prefix))
+	}
+	switch cfg.Compression {
+	case "gzip", "snappy", "":
```
[the remark this answers is not in the document]
Comment (bboreham) [reply pkg/util/grpcclient/grpcclient.go:58]: So this made me go back and look why I didn't use `gzip.Name`, and it turns out it didn't exist at the time.

File: pkg/util/grpc/encoding/snappy/snappy.go (hunk 6 above)
Comment (Wing924) [reply pkg/util/grpc/encoding/snappy/snappy.go:1]: Importing this package have a side effect: `encoding.RegisterCompressor(newCompressor())`
So I prefer not having  it directly in `pkg/util/grpc/encoding`

File: pkg/util/grpc/encoding/snappy/snappy.go (hunk 10 above)
Comment (Wing924) [reply pkg/util/grpc/encoding/snappy/snappy.go:18]: I've added tests.

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+type reader struct {
+	*snappy.Reader
+	pool *sync.Pool
+}
+
+func (r *reader) Read(p []byte) (n int, err error) {
+	n, err = r.Reader.Read(p)
+	if err == io.EOF {
+		r.pool.Put(r)
```
Comment (pstibrany) [inline pkg/util/grpc/encoding/snappy/snappy.go:78]: Nit: should we reset `snappy.Reader` with `nil` reader before putting back to pool? (same for writer) It would be little more correct to avoid keeping reference to just-read underlying reader, but likely not a big problem as they will be reused all the time.

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	pool *sync.Pool
+}
+
+func (w *writeCloser) Close() error {
+	defer w.pool.Put(w)
+	return w.Writer.Close()
+}
+
+type reader struct {
+	*snappy.Reader
```
Comment (pstibrany) [inline pkg/util/grpc/encoding/snappy/snappy.go:71]: Nit: These anonymous embeds now expose `snappy.Reader` methods. It doesn't even bring anything, as `Read` is fully reimplemented by `reader`. In `writeCloser` it saves reimplementing Write method, although I would still prefer to be explicit. Not a blocker for merging.

File: docs/configuration/config-file-reference.md [hunk 14]
```diff
@@ -1774,10 +1774,15 @@ bigtable:
...
 
-    # Use compression when sending messages.
+    # Deprecated: Use gzip compression when sending messages.
     # CLI flag: -bigtable.grpc-use-gzip-compression
     [use_gzip_compression: <boolean> | default = false]
 
+    # Use compression when sending messages. Supported values are: 'gzip',
+    # 'snappy' and '' (disable compression)
+    # CLI flag: -bigtable.grpc-compression
+    [compression: <string> | default = ""]
```
Comment (pracucci) [inline docs/configuration/config-file-reference.md:1784]: I would expect this to be `grpc_compression`. I'm not sure why the CI linter passed. Could you run `make doc` again, please?

File: pkg/util/grpcclient/grpcclient.go [hunk 15]
```diff
@@ -31,21 +38,40 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+
 // CallOptions returns the config in terms of CallOptions.
 func (cfg *Config) CallOptions() []grpc.CallOption {
 	var opts []grpc.CallOption
 	opts = append(opts, grpc.MaxCallRecvMsgSize(cfg.MaxRecvMsgSize))
 	opts = append(opts, grpc.MaxCallSendMsgSize(cfg.MaxSendMsgSize))
+	compression := cfg.GRPCCompression
 	if cfg.UseGzipCompression {
-		opts = append(opts, grpc.UseCompressor("gzip"))
+		compression = "gzip"
```
Comment (pracucci) [inline pkg/util/grpcclient/grpcclient.go:71]:
```suggestion
		compression = gzip.Name
```

File: pkg/util/grpcclient/grpcclient.go [hunk 16]
```diff
@@ -15,6 +21,7 @@ type Config struct {
 	MaxRecvMsgSize     int     `yaml:"max_recv_msg_size"`
 	MaxSendMsgSize     int     `yaml:"max_send_msg_size"`
 	UseGzipCompression bool    `yaml:"use_gzip_compression"`
```
Comment (pracucci) [inline pkg/util/grpcclient/grpcclient.go:23]:
```suggestion
	UseGzipCompression bool    `yaml:"use_gzip_compression"` // TODO: Remove this deprecated option in v1.6.0.
```

File: docs/configuration/config-file-reference.md (hunk 14 above)
Comment (Wing924) [reply docs/configuration/config-file-reference.md]: as you may know, CI don't run.

Comment (pracucci) [reply docs/configuration/config-file-reference.md]: Oh, I haven't noticed! Can't understand why the CI is not running on this specific PR 👀

**Comment by pracucci** [comment]
> I don't know why, but CI is not running on this PR. I'm falling it as "request changes" because we need to get the CI pass before merging.

I've found the issue. Your PR description contained some characters `スクリーンショット`  which apparently cause the CircleCI webhook to return 400. I've removed them and force pushed to trigger the CI.

FYI, **I've rebased the last commit** to reword the commit message by 1 character and `git push --force` it in order to trigger the CI again.

File: pkg/util/grpc/encoding/snappy/snappy.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
+	dr := c.readersPool.Get().(*snappy.Reader)
+	dr.Reset(r)
+	return reader{dr, &c.readersPool}, nil
+}
+
+type writeCloser struct {
+	*snappy.Writer
```
Comment (pracucci) [inline pkg/util/grpc/encoding/snappy/snappy.go:55]: Should this be a named `writer *snappy.Writer` too, according to the previous feedback?

Comment (Wing924) [reply pkg/util/grpc/encoding/snappy/snappy.go:55]: If we use `writer *snappy.Writer`, we need to add `func (* writeCloser)Write(p []byte) (n int, err error)`.
In additional, this type is private, no one can access it outside.

Comment (pstibrany) [reply pkg/util/grpc/encoding/snappy/snappy.go:55]: Type is private, but methods it implements are not. Right now it exposes 4 methods (`Reset`, `Write`, `Flush`, `Close`), which is not clear from the code, and it's clear whether it was intended or not.

Comment (pstibrany) [reply pkg/util/grpc/encoding/snappy/snappy.go:55]: (`Close` is reimplemented, so that one is clear)

**Comment by pstibrany** [comment]
Thank you!
